blockchain/fastsync.go

```diff
@@ -213,38 +212,63 @@ func worker(cache *utxoCache, jobs <-chan []byte, results chan<- *result) {
 // downloadUtxoSet will attempt to connect to make an HTTP GET request to the
 // provided sources one at a time and download the UTXO set to the provided path.
 // If a proxy is provided it will use it for the HTTP connection.
-func downloadUtxoSet(sources []string, proxy *socks.Proxy, pth string) (string, error) {
-	var fileName string
-
-	grab.SetDefaultClientWithProxy(proxy)
-
-	for _, src := range sources {
-		log.Infof("Downloading UTXO set from %s", src)
-		err := retry.Do(
-			func() error {
-				resp, err := grab.Get(pth, src)
-				if err != nil {
-					return err
+func downloadUtxoSet(sources []string, proxy *socks.Proxy, destination string) (string, error) {
+	for _, source := range sources {
+		log.Infof("Downloading UTXO set from %s to %s ...", source, destination)
+		retries := 3
+		for retries > 0 {
+			retries--
+
+			client := grab.NewClientWithProxy(proxy)
+			request, _ := grab.NewRequest(destination, source)
+
+			// Begin download
+			response := client.Do(request)
+
+			if response.HTTPResponse != nil {
+				fmt.Printf("  %v\n", response.HTTPResponse.Status)
+			}
+
+			displayDownloadProgress(response)
+
+			// Error checking
+			if responseError := response.Err(); responseError != nil {
+				log.Errorf("Failed downloading UTXO set: %s", responseError.Error())
+				if strings.Contains(responseError.Error(), "connection refused") {
+					break
+				} else {
+					continue
 				}
+			}
 
-				fileName = resp.Filename
+			filename := response.Filename
+			log.Infof("\nDownload saved to %v", filename)
 
-				return nil
-			},
-			retry.Attempts(3),
-			retry.RetryIf(func(err error) bool {
-				return !strings.Contains(err.Error(), "connection refused")
-			}),
-		)
-
-		if err != nil {
-			log.Errorf("Failed downloading UTXO set: %s", err.Error())
-			continue
+			log.Info("UTXO download complete. Verifying integrity...")
+			return filename, nil
 		}
-
-		log.Info("UTXO download complete. Verifying integrity...")
-		return fileName, nil
 	}
+	return "", AssertError("All UTXO sources are unavailable")
+}
+
+// displayDownloadProgress will display progress text while a file is being downloaded
+// The text is displayed on the same line by utilising the '\r' character
+func displayDownloadProgress(resp *grab.Response) {
+	// Begin UI loop
+	timer := time.NewTicker(500 * time.Millisecond)
+	defer timer.Stop()
 
-	return "", AssertError("all UTXO sources are unavailable")
+	for {
+		select {
+		case <-timer.C:
+			fmt.Printf("\r  transferred %v / %v bytes (%.2f%%)",
```

Comment:
Shouldn't this have a trailing newline?

Reply:
That would depend upon whether it's preferable to have the download update appear on the same line or across multiple lines.

`\r` Makes the text update on the same line

The main problem with this is that the printout looks messed up whenever more text gets logged to console during the download ( as can be seen in the screenshot given in a comment above )

The other option if to remove `\r` and have the `\n` at the end which would produce this output:

<img width="439" alt="Screen Shot 2019-03-30 at 5 14 02 pm" src="https://user-images.githubusercontent.com/2781841/55439552-b34a1700-55f0-11e9-9e16-0c79979422e2.png">

I'm fine with whichever displayed appearance is decided upon and will adjust this PR accordingly.

Reply:
I think it is better to keep them on different lines. Then make sure we don't display for fractional percentages... maybe a line every 10%? That way you can see the progress but it doesn't flood the logs?

Reply:
That sounds better.

It would mean the download text would be updated roughly every 30 seconds or so for a one hour download duration

Reply:
Yep, depends on the speed of their connection. I think periodic status is plenty to know if the download is slow or stalled. =)

Reply:
From testing, updating every 10 seconds looks better than 30 seconds at the moment.

What about the formating of the text ?

`FastSync: Downloaded 363204608 / 3102076686 bytes (12%)` ?

Reply:
I think it would be nice to use megabytes as the notation. This is probably easier to read.

```
FastSync: Downloaded 3.2MB / 3000MB (12%)
```

But I am open to suggestions. Bytes are also ok, just think its harder to read.

Reply:
I think logging the message based on % complete rather than time makes more sense. The timings are going to very a LOT from server to server depending on the download speed. If we just dump out stats every 10% then I think that suits all use cases well.

Reply:
Yep.
MB is better than the current bytes
363.2MB as opposed to 363204608 bytes or 363,204,608

Reply:
Prefered decimal places:

`FastSync: Downloaded 1272.19MB / 3102.08MB (41%)`
or
`FastSync: Downloaded 1272.2MB / 3102.1MB (41%)`

or no decimal places ?
`FastSync: Downloaded 1272MB / 3102MB (41%)`


`FastSync: Downloaded 1272MB / 3102MB (41%)`
or
`FastSync: Downloaded 1272MB of 3102MB (41%)`

Reply:
`FastSync: Downloaded 1,424MB of 3,102MB (45%)`
`FastSync: Downloaded 1,427MB of 3,102MB (46%)`
`FastSync: Downloaded 1,458MB of 3,102MB (47%)`
...

Reply:
```
FastSync: Downloaded 35MB of 3,102MB (1%)
FastSync: Downloaded 67MB of 3,102MB (2%)
FastSync: Downloaded 97MB of 3,102MB (3%)
FastSync: Downloaded 127MB of 3,102MB (4%)
```